### ISSUE TITLE
Endianness issues with airdecap-ng

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#1021]
**Reported by clopez on 26 Jul 2012 01:17 UTC**

When decrypting a .cap file with airdecap-ng the final result is different when you run it on a big endian architecture (MIPS) than when you run it on a little endian one (AMD64)

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:19 UTC**

Uploaded file [wpa-dec.bigendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-011948_000-wpa-dec.bigendian.cap)

debian-mips ~ # airdecap-ng -e test -p biscotte test/wpa.cap

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:20 UTC**

Uploaded file [wpa-dec.littleendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-012013_000-wpa-dec.littleendian.cap)

debian-amd64 ~ # airdecap-ng -e test -p biscotte test/wpa.cap

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

Attached two files.

Both were generated running 

```
airdecap-ng -e test -p biscotte test/wpa.cap
```
test/wpa.cap is the file that comes on the test/ directory of aircrack-ng

Compare both files with some bindiff tool and you will see that the bytes are reversed on some parts of the file

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:26 UTC**

Uploaded file [wpa-psk-linksys-dec.bigendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-012655_000-wpa-psk-linksys-dec.bigendian.cap)

debian-mips ~ # airdecap-ng -e linksys -p dictionary test/wpa-psk-linksys.cap

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:27 UTC**

Uploaded file [wpa-psk-linksys-dec.littleendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-012718_000-wpa-psk-linksys-dec.littleendian.cap)

debian-amd64 ~ # airdecap-ng -e linksys -p dictionary test/wpa-psk-linksys.cap

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

Attached two more files.

Both were generated running

```
airdecap-ng -e linksys -p dictionary test/wpa-psk-linksys.cap
```

test/wpa-psk-linksys.cap is the file that comes on the test/ directory of aircrack-ng (after aircrack-ng/aircrack-ng@1bd45dddf8578e0c477fb7c38c34e8aabacdfa7b)

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:28 UTC**

Uploaded file [wpa2-psk-linksys-dec.bigendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-012851_000-wpa2-psk-linksys-dec.bigendian.cap)

debian-mips ~ # airdecap-ng -e linksys -p dictionary test/wpa2-psk-linksys.cap

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 26 Jul 2012 01:29 UTC**

Uploaded file [wpa2-psk-linksys-dec.littleendian.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20120726-012912_000-wpa2-psk-linksys-dec.littleendian.cap)

debian-amd64 ~ # airdecap-ng -e linksys -p dictionary test/wpa2-psk-linksys.cap

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

And attached two more files.

Both were generated running

```
airdecap-ng -e linksys -p dictionary test/wpa2-psk-linksys.cap
```

test/wpa2-psk-linksys.cap is the file that comes on the test/ directory of aircrack-ng (after aircrack-ng/aircrack-ng@1bd45dddf8578e0c477fb7c38c34e8aabacdfa7b)

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

Note that on the last two cases (wpa-psk-linksys-dec.cap and wpa2-psk-linksys-dec.cap) the files generated on the big endian machine are completely broken:

```
$ hexdump wpa-psk-linksys-dec.bigendian.cap 
0000000 b2a1 d4c3 0200 0400 0000 0000 0000 0000
0000010 0000 ffff 0000 0100                    
0000018

$ hexdump wpa2-psk-linksys-dec.bigendian.cap 
0000000 b2a1 d4c3 0200 0400 0000 0000 0000 0000
0000010 0000 ffff 0000 0100                    
0000018
```

Seems that airdecap-ng is not recognizing the wpa packets
```
debian-mips aircrack-ng # ./src/airdecap-ng -e linksys -p dictionary test/wpa-psk-linksys.cap 
Total number of packets read           587
Total number of WEP data packets         0
Total number of WPA data packets        57
Number of plaintext data packets         0
Number of decrypted WEP  packets         0
Number of corrupted WEP  packets         0
Number of decrypted WPA  packets         0

debian-mips aircrack-ng # ./src/airdecap-ng -e linksys -p dictionary test/wpa2-psk-linksys.cap 
Total number of packets read           499
Total number of WEP data packets         0
Total number of WPA data packets        28
Number of plaintext data packets         0
Number of decrypted WEP  packets         0
Number of corrupted WEP  packets         0
Number of decrypted WPA  packets         0
```

All tests were  done and packets were generated with the last svn version of aircrack-ng/airdecap-ng (aircrack-ng/aircrack-ng@ff7dc7297ce544de01481730c36c816b832cd9ad)

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 27 Feb 2013 01:57 UTC**

Uploaded file [airdecap-ng.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20130227-015728_000-airdecap-ng.patch)

Patch for endianness issues on airdecap-ng

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

I have ported the patch from #1022 to airdecap-ng.c (see it attached above)


With this patch all works as expected. The hash of the file is still different. But now at least works.

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 27 Feb 2013 01:59 UTC**

Uploaded file [wpa-psk-linksys-dec.bigendian.fixed.cap](https://github.com/aircrack-ng/trac-attachments/blob/master/1021/20130227-015938_000-wpa-psk-linksys-dec.bigendian.fixed.cap)

After applying patch: ./src/airdecap-ng -e linksys -p dictionary test/wpa-psk-linksys.cap

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

The file attached above is what now airdecap-ng produces after patching it.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 26 Jul 2012 01:17 UTC**

ACKed.

**Comment by aircrack-ng** [comment]
**Comment by clopez on 26 Jul 2012 01:17 UTC**

Fixed on aircrack-ng/aircrack-ng@2a0a6b32fdf26aea995c44f31e9abd5789c4f798
